javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,80 +14,187 @@ package java.util
-    this(new js.Array[E])
+    this(
+      if (isWebAssembly)
+        new Array[Any](Math.max(initialCapacity, 0))
+      else new js.Array[E],
+      0
+    )
     if (initialCapacity < 0)
       throw new IllegalArgumentException
   }
 
   def this() =
-    this(new js.Array[E])
+    this(
+      if (isWebAssembly) new Array[Any](16)
+      else new js.Array[E],
+      0
+    )
 
   def this(c: Collection[_ <: E]) = {
-    this()
+    this(
+      if (isWebAssembly) new Array[Any](c.size())
+      else new js.Array[E],
+      0
+    )
     addAll(c)
   }
 
   def trimToSize(): Unit = {
-    // We ignore this as js.Array doesn't support explicit pre-allocation
+    if (isWebAssembly) resizeTo(size())
+    // We ignore this in JS as js.Array doesn't support explicit pre-allocation
   }
 
   def ensureCapacity(minCapacity: Int): Unit = {
-    // We ignore this as js.Array doesn't support explicit pre-allocation
+    if (isWebAssembly) {
+      if (innerWasm.length < minCapacity)
+        resizeTo(minCapacity)
+    }
+    // We ignore this in JS as js.Array doesn't support explicit pre-allocation
   }
 
   def size(): Int =
-    inner.length
-
-  override def clone(): AnyRef =
-    new ArrayList(inner.jsSlice(0))
+    if (isWebAssembly) _size
+    else innerJS.length
+
+  override def clone(): AnyRef = {
+    if (isWebAssembly) {
+      new ArrayList(innerWasm.clone(), size())
+    } else {
+      new ArrayList(innerJS.jsSlice(0), 0)
+    }
+  }
 
   def get(index: Int): E = {
     checkIndexInBounds(index)
-    inner(index)
+    if (isWebAssembly) {
+      innerWasm(index).asInstanceOf[E]
+    } else {
+      innerJS(index)
+    }
   }
 
   override def set(index: Int, element: E): E = {
     val e = get(index)
-    inner(index) = element
+    if (isWebAssembly) {
+      innerWasm(index) = element
+    } else {
+      innerJS(index) = element
+    }
     e
   }
 
   override def add(e: E): Boolean = {
-    inner.push(e)
+    if (isWebAssembly) {
+      add(_size, e)
+    } else {
+      innerJS.push(e)
+    }
     true
   }
 
   override def add(index: Int, element: E): Unit = {
     checkIndexOnBounds(index)
-    inner.splice(index, 0, element)
+    if (isWebAssembly) {
+      if (size() >= innerWasm.length) expand()
+      System.arraycopy(innerWasm, index, innerWasm, index + 1, size() - index)
+      innerWasm(index) = element
+      _size += 1
+    } else {
+      innerJS.splice(index, 0, element)
+    }
   }
 
   override def remove(index: Int): E = {
     checkIndexInBounds(index)
-    arrayRemoveAndGet(inner, index)
+    if (isWebAssembly) {
+      val removed = innerWasm(index).asInstanceOf[E]
+      System.arraycopy(innerWasm, index + 1, innerWasm, index, size() - index - 1)
+      innerWasm(size - 1) = null
+      _size -= 1
+      removed
+    } else {
+      arrayRemoveAndGet(innerJS, index)
+    }
   }
 
   override def clear(): Unit =
-    inner.length = 0
+    if (isWebAssembly) {
+      removeRange(0, size())
```

Comment:
This never needs to check bounds nor move elements. Consider using a direct `Arrays.fill`:
```suggestion
      Arrays.fill(innerWasm, null) // free references for GC
```